### PR TITLE
[Tables] Add multitenant challenge auth policy support

### PR DESCRIPTION
**Pull request by mccoyp**
# Description

See https://github.com/Azure/azure-sdk-for-python/pull/24019 for details about the included policy, and multitenant challenge authentication more generally. This adds the proposed policy to Tables for now, while a formal review is pending for moving this into Core.

To enable 401 responses that include auth challenges from the service, this adds support for API version `2020-12-06`. Earlier API versions trigger 403 responses upon failed authentication, so they bypass the `on_challenge` method of the new policy and have an auth flow identical to that of the `BearerTokenCredentialPolicy` that's used today.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.
  - In addition to the included tests, both the sync and async policies have been manually tested for multitenant and token caching behavior for new and old API versions.

### CODE REVIEW

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py [hunk 1]
```diff
@@ -249,7 +248,7 @@ def _configure_policies(self, **kwargs):
     def _configure_credential(self, credential):
         # type: (Any) -> None
         if hasattr(credential, "get_token"):
-            self._credential_policy = BearerTokenCredentialPolicy(  # type: ignore
+            self._credential_policy = BearerTokenChallengePolicy(  # type: ignore
                 credential, STORAGE_OAUTH_SCOPE
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:252]: So tables uses the default `True` for the boolean flags?

File: sdk/tables/azure-data-tables/tests/test_challenge_auth.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+from azure.core.pipeline import Pipeline
+from azure.core.pipeline.transport import HttpRequest as PipelineTransportHttpRequest
+from azure.core.rest import HttpRequest as RestHttpRequest
+from azure.data.tables._authentication import BearerTokenChallengePolicy
+import pytest
+
+HTTP_REQUESTS = [PipelineTransportHttpRequest, RestHttpRequest]
+
+
+@pytest.mark.parametrize("http_request", HTTP_REQUESTS)
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/tests/test_challenge_auth.py:19]: Is it possible to maybe add a couple of 'playback only' recorded tests?

Comment (mccoyp) [reply sdk/tables/azure-data-tables/tests/test_challenge_auth.py:78]: Added!

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py (hunk 1 above)
Comment (mccoyp) [reply sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:252]: It does -- this works for API versions that support auth challenges as well as those that don't. Tables challenges contain a valid `resource` field in the WWW-Authenticate header, so scopes discovery does work correctly

File: sdk/tables/azure-data-tables/tests/test_challenge_auth_async.py [hunk 3]
```diff
@@ -7,15 +7,74 @@
...
+    from azure.core.rest import HttpRequest as RestHttpRequest
+    HTTP_REQUESTS = [PipelineTransportHttpRequest, RestHttpRequest]
+except ModuleNotFoundError:
+    HTTP_REQUESTS = [PipelineTransportHttpRequest]
+
+
+class TestTableChallengeAuthAsync(AzureRecordedTestCase, AsyncTableTestCase):
+    @tables_decorator_async
+    @recorded_by_proxy_async
+    async def test_challenge_auth_supported_version(self, tables_storage_account_name):
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/tests/test_challenge_auth_async.py:35]: These tests will be skipped with the CI live runs?

**Review by annatisch — APPROVED** [review]

File: sdk/tables/azure-data-tables/tests/test_challenge_auth_async.py (hunk 3 above)
Comment (mccoyp) [reply sdk/tables/azure-data-tables/tests/test_challenge_auth_async.py:35]: Yep, that's correct. The default settings for `New-TestResources` service principals don't enable multitenant authentication, so there's no easy way to automate these tests in live mode (that I know of)

Comment (annatisch) [reply sdk/tables/azure-data-tables/tests/test_challenge_auth_async.py:35]: Perfect - thanks!

**Pull request merged**